councilmatic_core/management/commands/convert_attachment_text.py

```diff
@@ -0,0 +1,111 @@
+
+    def handle(self, *args, **options):
+        self.update_all = options['update_all']
+        self.add_plain_text()
+
+    def get_document_url(self):
+        with engine.begin() as connection:
+            # Only apply this query to most recently updated (or created) bill documents.
+            max_updated = BillDocument.objects.all().aggregate(Max('updated_at'))['updated_at__max']
+
+            if max_updated is None or self.update_all:
+                query = '''
+                    SELECT id, url
+                    FROM councilmatic_core_billdocument 
+                    WHERE document_type='A' 
+                    AND full_text is null
+                    AND lower(url) similar to '%(.doc|.docx|.pdf)'
+                    ORDER BY updated_at DESC
+                ''' 
+            else:
+                query = '''
+                    SELECT id, url
+                    FROM councilmatic_core_billdocument 
+                    WHERE updated_at >= :max_updated
+                    AND document_type='A' 
+                    AND full_text is null
+                    AND lower(url) similar to '%(.doc|.docx|.pdf)'
+                    ORDER BY updated_at DESC
+                '''
+
+            result = connection.execution_options(stream_results=True).execute(sa.text(query), max_updated=max_updated)
+
+            yield from result
+
+    def convert_document_to_plaintext(self):
+        logger.info('Converting document to plain text...')
+
+        for document_data in self.get_document_url():
+            document_data = dict(document_data)
+            url = document_data['url']
+            document_id = document_data['id']
+            response = requests.get(url)
+            extension = os.path.splitext(url)[1]
+
+            with tempfile.NamedTemporaryFile(suffix=extension) as tfp:
+                tfp.write(response.content)
+                plain_text = textract.process(tfp.name)
+
+            yield {'plain_text': plain_text.decode('utf-8'), 'id': document_id}
+            
+
+    def add_plain_text(self):
+        '''
+        Metro has over 2,000 attachments that should be converted into plain text. 
+        Updating all attachments in a single query can be memory intensive.
+        This function insures that the database updates only 20 documents per connection: it fetches up to 20 elements from a generator object, runs the UPDATE query, and then fetches up to 20 more.
+
+        Inspired by: https://stackoverflow.com/questions/30510593/how-can-i-use-server-side-cursors-with-django-and-psycopg2/41088159#41088159
+        '''
+        update_statement = '''
+            UPDATE councilmatic_core_billdocument as bill_docs
+            SET full_text = :plain_text
+            WHERE bill_docs.id = :id  
+        '''
+        
+        plaintexts = self.convert_document_to_plaintext()
+
+        while True:
+            plaintexts_fetched_from_generator = list(itertools.islice(plaintexts, 20))
```

Comment:
If @fgregg wants to backread this change just in case I've missed a subtlety, I'd be much obliged!

Reply:
Sounds good @hancush ! I'll wait for final comments from @fgregg before merging.